File: lib/worldbank_proxy.py
"""World Bank Open Data — 군사비 통계 (키 불필요)"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from datetime import datetime, timezone

WB_BASE = "https://api.worldbank.org/v2"

GDP_COUNTRIES = [
    ("USA", "미국"),
    ("SAU", "사우디"),
    ("ISR", "이스라엘"),
    ("RUS", "러시아"),
    ("KOR", "한국"),
    ("FRA", "프랑스"),
    ("GBR", "영국"),
    ("CHN", "중국"),
    ("JPN", "일본"),
    ("DEU", "독일"),
]

REGION_COUNTRIES = {
    "북미": ["USA", "CAN"],
    "유럽": ["DEU", "FRA", "GBR", "ITA", "ESP", "POL", "NLD", "NOR", "SWE", "UKR"],
    "아시아·태평양": ["CHN", "JPN", "KOR", "IND", "AUS", "TWN", "VNM", "IDN"],
    "중동": ["SAU", "ISR", "IRN", "ARE", "TUR", "EGY"],
}

REGION_COLORS = {
    "북미": ("#4da6ff", "rgba(77, 166, 255, 0.1)"),
    "유럽": ("#b48cff", "rgba(180, 140, 255, 0.08)"),
    "아시아·태평양": ("#3dd68c", "rgba(61, 214, 140, 0.08)"),
    "중동": ("#f0a030", "rgba(240, 160, 48, 0.08)"),
}

ALL_COUNTRY_CODES = sorted({
    *{c for c, _ in GDP_COUNTRIES},
    *{c for codes in REGION_COUNTRIES.values() for c in codes},
    "WLD",
    "KOR",
})


def _fetch_indicator(codes: list[str], indicator: str, date_range: str = "2019:2023") -> list[dict]:
    chunk_size = 12
    rows: list[dict] = []
    for i in range(0, len(codes), chunk_size):
        chunk = codes[i : i + chunk_size]
        url = f"{WB_BASE}/country/{';'.join(chunk)}/indicator/{indicator}?format=json&per_page=2000&date={date_range}"
        req = urllib.request.Request(url, headers={"User-Agent": "LIG-Dashboard/1.0 (worldbank-proxy)"})
        with urllib.request.urlopen(req, timeout=45) as resp:
            data = json.loads(resp.read().decode("utf-8"))
        chunk_rows = data[1] if isinstance(data, list) and len(data) > 1 else []
        rows.extend(r for r in chunk_rows if r.get("value") is not None)
    return rows


def _group(rows: list[dict]) -> dict[str, dict[str, float]]:
    out: dict[str, dict[str, float]] = {}
    for row in rows:
        code = row["countryiso3code"]
        out.setdefault(code, {})[row["date"]] = row["value"]
    return out


def _sorted_years(by_country: dict, codes: list[str]) -> list[str]:
    years: set[str] = set()
    for code in codes:
        years.update(by_country.get(code, {}).keys())
    return sorted(years)


def fetch_worldbank_stats() -> dict:
    usd_rows = _fetch_indicator(ALL_COUNTRY_CODES, "MS.MIL.XPND.CD")
    gdp_rows = _fetch_indicator([c for c, _ in GDP_COUNTRIES] + ["WLD"], "MS.MIL.XPND.GD.ZS")

    by_usd = _group(usd_rows)
    by_gdp = _group(gdp_rows)
    years = _sorted_years(by_usd, ALL_COUNTRY_CODES)

    regional_datasets = []
    for region, codes in REGION_COUNTRIES.items():
        border, bg = REGION_COLORS[region]
        regional_datasets.append({
            "label": region,
            "data": [
                round(sum(by_usd.get(c, {}).get(y, 0) for c in codes) / 1e9, 1)
                for y in years
            ],
            "borderColor": border,
            "backgroundColor": bg,
            "tension": 0.3,
            "fill": True,
        })

    gdp_year = _sorted_years(by_gdp, [c for c, _ in GDP_COUNTRIES])[-1]
    gdp_items = [
        {"label": label, "value": by_gdp.get(code, {}).get(gdp_year)}
        for code, label in GDP_COUNTRIES
        if by_gdp.get(code, {}).get(gdp_year) is not None
    ]
    gdp_items.sort(key=lambda x: x["value"], reverse=True)

    kor = by_usd.get("KOR", {})
    kor_years = sorted(kor.keys())

    world = by_usd.get("WLD", {})
    w_years = sorted(world.keys())
    latest = w_years[-1]
    prev = w_years[-2] if len(w_years) > 1 else latest
    latest_val = world[latest]
    prev_val = world[prev]
    yoy = round(((latest_val - prev_val) / prev_val) * 100, 1) if prev_val else 0

    top5 = ["USA", "CHN", "RUS", "IND", "SAU"]
    top5_spend = sum(by_usd.get(c, {}).get(latest, 0) for c in top5)
    top5_share = round((top5_spend / latest_val) * 100, 1) if latest_val else None

    kor_gdp = by_gdp.get("KOR", {})
    kor_gdp_years = sorted(kor_gdp.keys())
    kor_latest = kor_gdp_years[-1]
    kor_prev = kor_gdp_years[-2] if len(kor_gdp_years) > 1 else kor_latest
    kor_val = kor_gdp.get(kor_latest)
    kor_change = round(kor_val - kor_gdp.get(kor_prev, kor_val), 1) if kor_val is not None else None

    return {
        "updated": datetime.now(timezone.utc).isoformat(),
        "source": "World Bank Open Data",
        "live": True,
        "regional": {
            "labels": years,
            "datasets": regional_datasets,
            "unit": "십억 USD",
            "source": "World Bank",
        },
        "gdpRatio": {
            "labels": [i["label"] for i in gdp_items],
            "data": [round(i["value"], 1) for i in gdp_items],
            "year": gdp_year,
            "source": "World Bank",
        },
        "koreaSpending": {
            "labels": kor_years,
            "data": [round(kor[y] / 1e9, 1) for y in kor_years],
            "unit": "십억 USD",
            "source": "World Bank",
            "note": "국방비 지출 (방산 수출액과 다름)",
        },
        "kpis": {
            "globalSpending": {
                "value": latest_val,
                "formatted": f"${latest_val / 1e12:.2f}T",
                "year": latest,
                "yoy": yoy,
                "source": "World Bank",
            },
            "top5Share": {
                "value": top5_share,
                "formatted": f"{top5_share}%" if top5_share is not None else "—",
                "year": latest,
                "note": "군사비 지출 상위 5개국 점유율",
                "source": "World Bank",
            },
            "koreaGdpRatio": {
                "value": kor_val,
                "formatted": f"{kor_val:.1f}%" if kor_val is not None else "—",
                "year": kor_latest,
                "change": kor_change,
                "source": "World Bank",
            },
        },
    }

File: lib/worldbank-proxy.js
/**
 * World Bank Open Data — 군사비 지출 (키 불필요)
 * https://data.worldbank.org/indicator/MS.MIL.XPND.CD
 */

const WB_BASE = "https://api.worldbank.org/v2";

let statsCache = null;
let statsCacheAt = 0;
const STATS_CACHE_MS = 15 * 60 * 1000;

const GDP_COUNTRIES = [
  { code: "USA", label: "미국" },
  { code: "SAU", label: "사우디" },
  { code: "ISR", label: "이스라엘" },
  { code: "RUS", label: "러시아" },
  { code: "KOR", label: "한국" },
  { code: "FRA", label: "프랑스" },
  { code: "GBR", label: "영국" },
  { code: "CHN", label: "중국" },
  { code: "JPN", label: "일본" },
  { code: "DEU", label: "독일" },
];

const REGION_COUNTRIES = {
  "북미": ["USA", "CAN"],
  "유럽": ["DEU", "FRA", "GBR", "ITA", "ESP", "POL", "NLD", "NOR", "SWE", "UKR"],
  "아시아·태평양": ["CHN", "JPN", "KOR", "IND", "AUS", "TWN", "VNM", "IDN"],
  "중동": ["SAU", "ISR", "IRN", "ARE", "TUR", "EGY"],
};

const REGION_COLORS = {
  "북미": { border: "#4da6ff", bg: "rgba(77, 166, 255, 0.1)" },
  "유럽": { border: "#b48cff", bg: "rgba(180, 140, 255, 0.08)" },
  "아시아·태평양": { border: "#3dd68c", bg: "rgba(61, 214, 140, 0.08)" },
  "중동": { border: "#f0a030", bg: "rgba(240, 160, 48, 0.08)" },
};

const ALL_COUNTRY_CODES = [
  ...new Set([
    ...GDP_COUNTRIES.map((c) => c.code),
    ...Object.values(REGION_COUNTRIES).flat(),
    "WLD",
    "KOR",
  ]),
];

async function fetchWbIndicator(countryCodes, indicator, dateRange = "2019:2023") {
  const chunkSize = 12;
  const rows = [];
  for (let i = 0; i < countryCodes.length; i += chunkSize) {
    const chunk = countryCodes.slice(i, i + chunkSize);
    const codes = chunk.join(";");
    const url =
      `${WB_BASE}/country/${codes}/indicator/${indicator}` +
      `?format=json&per_page=2000&date=${dateRange}`;

    const res = await fetch(url, {
      headers: { "User-Agent": "LIG-Dashboard/1.0 (worldbank-proxy)" },
    });
    if (!res.ok) throw new Error(`World Bank HTTP ${res.status}`);

    const json = await res.json();
    const part = Array.isArray(json) ? json[1] : null;
    if (!Array.isArray(part)) throw new Error("World Bank response invalid");
    rows.push(...part.filter((r) => r.value != null));
  }
  return rows;
}

function groupByCountryYear(rows) {
  const map = {};
  for (const row of rows) {
    const code = row.countryiso3code;
    if (!map[code]) map[code] = {};
    map[code][row.date] = row.value;
  }
  return map;
}

function sortedYears(map, codes) {
  const years = new Set();
  for (const code of codes) {
    Object.keys(map[code] || {}).forEach((y) => years.add(y));
  }
  return [...years].sort();
}

function buildRegionalChart(usdRows) {
  const byCountry = groupByCountryYear(usdRows);
  const years = sortedYears(byCountry, ALL_COUNTRY_CODES);

  const datasets = Object.entries(REGION_COUNTRIES).map(([region, codes]) => ({
    label: region,
    data: years.map((year) => {
      const sum = codes.reduce((acc, code) => acc + (byCountry[code]?.[year] || 0), 0);
      return Math.round((sum / 1e9) * 10) / 10;
    }),
    borderColor: REGION_COLORS[region].border,
    backgroundColor: REGION_COLORS[region].bg,
    tension: 0.3,
    fill: true,
  }));

  return { labels: years, datasets, unit: "십억 USD", source: "World Bank" };
}

function buildGdpRatioChart(gdpRows) {
  const byCountry = groupByCountryYear(gdpRows);
  const latestYear = sortedYears(byCountry, GDP_COUNTRIES.map((c) => c.code)).pop();

  const items = GDP_COUNTRIES.map(({ code, label }) => ({
    label,
    value: byCountry[code]?.[latestYear] ?? null,
  }))
    .filter((i) => i.value != null)
    .sort((a, b) => b.value - a.value);

  return {
    labels: items.map((i) => i.label),
    data: items.map((i) => Math.round(i.value * 10) / 10),
    year: latestYear,
    source: "World Bank",
  };
}

function buildKoreaSpendingChart(usdRows) {
  const byCountry = groupByCountryYear(usdRows);
  const korea = byCountry.KOR || {};
  const years = Object.keys(korea).sort();
  const data = years.map((y) => Math.round((korea[y] / 1e9) * 10) / 10);

  return {
    labels: years,
    data,
    unit: "십억 USD",
    source: "World Bank",
    note: "국방비 지출 (방산 수출액과 다름)",
  };
}

function buildKpis(usdRows, gdpRows) {
  const byCountry = groupByCountryYear(usdRows);
  const world = byCountry.WLD || {};
  const years = Object.keys(world).sort();
  const latest = years[years.length - 1];
  const prev = years[years.length - 2];
  const latestVal = world[latest];
  const prevVal = world[prev];
  const yoy = prevVal ? ((latestVal - prevVal) / prevVal) * 100 : 0;

  const gdpByCountry = groupByCountryYear(gdpRows);
  const gdpYear = sortedYears(gdpByCountry, GDP_COUNTRIES.map((c) => c.code)).pop();

  const top5Spenders = ["USA", "CHN", "RUS", "IND", "SAU"];
  const top5Spend = top5Spenders.reduce((s, c) => s + (byCountry[c]?.[latest] || 0), 0);
  const top5Share = latestVal ? (top5Spend / latestVal) * 100 : null;

  const korGdp = gdpByCountry.KOR || {};
  const korYears = Object.keys(korGdp).sort();
  const korLatest = korYears[korYears.length - 1];
  const korPrev = korYears[korYears.length - 2];
  const korGdpVal = korGdp[korLatest];
  const korGdpChange = korGdpVal != null && korGdp[korPrev] != null
    ? korGdpVal - korGdp[korPrev]
    : null;

  return {
    globalSpending: {
      value: latestVal,
      formatted: `$${(latestVal / 1e12).toFixed(2)}T`,
      year: latest,
      yoy: Math.round(yoy * 10) / 10,
      source: "World Bank",
    },
    top5Share: {
      value: top5Share,
      formatted: top5Share != null ? `${top5Share.toFixed(1)}%` : "—",
      year: latest,
      note: "군사비 지출 상위 5개국 점유율",
      source: "World Bank",
    },
    koreaGdpRatio: {
      value: korGdpVal,
      formatted: korGdpVal != null ? `${korGdpVal.toFixed(1)}%` : "—",
      year: korLatest,
      change: korGdpChange,
      source: "World Bank",
    },
  };
}

async function fetchWorldBankStats() {
  if (statsCache && Date.now() - statsCacheAt < STATS_CACHE_MS) {
    return statsCache;
  }

  const [usdRows, gdpRows] = await Promise.all([
    fetchWbIndicator(ALL_COUNTRY_CODES, "MS.MIL.XPND.CD", "2019:2023"),
    fetchWbIndicator([...GDP_COUNTRIES.map((c) => c.code), "WLD"], "MS.MIL.XPND.GD.ZS", "2019:2023"),
  ]);

  statsCache = {
    updated: new Date().toISOString(),
    source: "World Bank Open Data",
    live: true,
    regional: buildRegionalChart(usdRows),
    gdpRatio: buildGdpRatioChart(gdpRows),
    koreaSpending: buildKoreaSpendingChart(usdRows),
    kpis: buildKpis(usdRows, gdpRows),
  };
  statsCacheAt = Date.now();
  return statsCache;
}

module.exports = {
  fetchWorldBankStats,
  GDP_COUNTRIES,
  REGION_COUNTRIES,
};
